### ISSUE TITLE
v0.25.2 doesn't reliably run preload scripts in Webviews

### ISSUE THREAD

**Issue by paulcbetts** [atom/electron#1560]
Our last big blocker for moving to Chrome 42 is that we're seeing the preload script not execute reliably - on initial load it doesn't execute at all, and refreshing the page will allow some of them to execute but not all of them.

I'll dig into `renderer/lib/init.coffee` to see if I can get more actionable information than "It doesn't work Halp!"

**Comment by zcbenz** [comment]
The command line parameters of webview's renderer process would be very helpful, you can get it via `process.argv` in webview's devtools.

**Comment by paulcbetts** [comment]
Hm, more weirdness - this is almost a 100% repro on OS X, but on Windows I haven't seen it yet. Here's `process.argv` on OS X, sure enough the preload script isn't there:

```
process.argv
["/private/var/folders/_c/hwfdpfg168s0tl7dlqzfrf_h0000gn/T/slack-build/Slack.app/Contents/Frameworks/Slack Helper.app/Contents/MacOS/Slack Helper", "--type=renderer", "--js-flags=--harmony", "--no-sandbox", "--enable-deferred-image-decoding", "--lang=English", "--enable-pinch-virtual-viewport", "--enable-delegated-renderer", "--num-raster-threads=2", "--channel=60816.7.559197518"]
```

**Comment by paulcbetts** [comment]
I take it back, it's appearing in the login screen periodically, here's a Windows example:

>  --type=renderer --js-flags=--harmony --no-sandbox --enable-deferred-image-decoding --lang=en-GB --enable-pinch --device-scale-factor=1 --font-cache-shared-mem-suffix=2376 --enable-pinch-virtual-viewport --enable-delegated-renderer --num-raster-threads=1 --channel="2376.5.1597965734\645508156" /prefetch:673131151

Strangely enough, I'm getting more render processes than I'm expecting - with a single WebView on screen:

![](http://cl.ly/image/0a203g2V210w/content#png)

**Comment by zcbenz** [comment]
From the command line parameters Electron failed to append custom switches for webview's renderer process, can you work out a reproducible example? The simplest code seems to work:

```js
var w = new WebView;
w.setAttribute('preload', 'file:///tmp/test.js');
w.src = 'http://google.com';
document.body.appendChild(w);
```

**Comment by paulcbetts** [comment]
@zcbenz What happens if you do like 10 of those in a row? It seems to be timing-related, because I can affect how likely it will happen with opening devtools at the right time

**Comment by paulcbetts** [comment]
@zcbenz I'll work on getting a repro for this tomorrow

**Comment by zcbenz** [comment]
Openning 10 webview in the same time works for me.

**Comment by paulcbetts** [comment]
Ok! So it _appears_ that this is related somehow to setting the user agent at the wrong time. I'm trying to get a failing test to prove it

**Comment by paulcbetts** [comment]
It seems like a super reliable repro is:

```
var w = new WebView;
w.setAttribute('preload', 'file:///tmp/test.js');
w.src = 'http://google.com';
w.addEventListener('did-start-load', function() { w.setUserAgent("CustomUserAgent 1.0.0") });
document.body.appendChild(w);
```

It's difficult to find a way to call `setUserAgent` before the page loads correctly - calling it immediately will throw an exception, call it too late and it doesn't get applied. The only way that I've reliably found is to send an IPC message from the very top of the preload script to the renderer.

**Comment by paulcbetts** [comment]
This is fixed by #1587
